[Docs] Add api gateways documentation (#2617)
nuclio/nuclio: docs/reference/api-gateway/http.md

> # API Gateway with http requests

Reviewer: there's only apigateway for http requests, so perhaps no need to explicitly specify that

#### In this document

> - [Without Authentication](#none-auth)

Reviewer: Change to "No authentication" as it express better there is no authentication ;)

    - [Create](#create-none)
    - [Invoke](#invoke-none)
- [Basic Authentication](#basic-auth)
    - [Create](#create-basic)
    - [Invoke](#invoke-basic)
- [Canary Function](#canary-function)

<a id="none-auth"></a>
## Without Authentication

<a id="create-none"></a>
### Create API Gateways

You can create an api gateway with basic authentication by sending a POST request to the following endpoint:

```
<nuclio-host-name>/api/api_gateways
```
With the following request body:
```json
{
    "spec": {
        "name": "<apigateway-name>",
        "description": "some description",
        "path": "/some/path", 
        "authenticationMode": "none",
        "upstreams": [
            {
                "kind": "nucliofunction",
                "nucliofunction": {
                    "name": "function-name-to-invoke",
                },
                "percentage": 0
            }
        ],
        "host": "<apigateway-name>-<project-name>.<nuclio-host-name>"
    },
    "metadata": {
        "labels": {
            "nuclio.io/project-name": "default"
        },
        "name": "<apigateway-name>"
    }
}
```

<a id="invoke-none"></a>
### Invoke API Gateways

> To invoke it, simply send GET/POST request to `<apigateway-name>-<project-name>.<nuclio-host-name>` (aka - `spec.host` )

Reviewer: ```suggestion
To invoke it, simply send a request to the created API Gateway ingress (e.g.: `<apigateway-name>-<project-name>.<nuclio-host-name>`, specified on the request body `spec.host` )
```

> If your function accepts a request body, you can send it as a request body.

Reviewer: irrelevant

<a id="basic-auth"></a>
## Basic Authentication

> Basic authentication is a way to authenticate users by providing a username and password.

Reviewer: ```suggestion
You may protect your function by setting the API Gateway with [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication). That means, to access a function, client would need to provide both username & password.

<a id="create-basic"></a>
### Create API Gateways

You can create an api gateway with basic authentication by sending a POST request to the following endpoint:

```
<nuclio-host-name>/api/api_gateways
```
With the following request body:
```json
{
    "spec": {
        "name": "<apigateway-name>",
        "description": "some description",
        "path": "/some/path",
        "authenticationMode": "basicAuth",
        "upstreams": [
            {
                "kind": "nucliofunction",
                "nucliofunction": {
                    "name": "function-name-to-invoke"
                },
                "percentage": 0
            }
        ],
        "host": "<apigateway-name>-<project-name>.<nuclio-host-name>",
        "authentication": {
            "basicAuth": {
                "username": "some-username",
                "password": "some-password"
            }
        }
    },
    "metadata": {
        "labels": {
            "nuclio.io/project-name": "default"
        },
        "name": "<apigateway-name>"
    }
}
```

<a id="invoke-basic"></a>
### Invoke API Gateways

> To invoke it, send GET/POST request to `<apigateway-name>-<project-name>.<nuclio-host-name>` (aka - `spec.host` )

Reviewer: ```suggestion
To invoke it, send GET/POST request to `<apigateway-name>-<project-name>.<nuclio-host-name>` (aka - `spec.host` )
```

see comment above

with the following header:
```
key: Authorization
value: Basic base64encode("username:password")
```

*Example*:

For the following credentials:
```
"username": "some-username"
"password": "some-password"
```

> base64 encoding of "some-username:some-password" is `c29tZS11c2VybmFtZTpzb21lLXBhc3N3b3Jk`, so the resulting header is:

Reviewer: ```suggestion
base64 encoding of "some-username:some-password" is `c29tZS11c2VybmFtZTpzb21lLXBhc3N3b3Jk` (`echo "some-username:some-password" | base64 -d`), so the resulting header is:
```

```
Authorization: Basic c29tZS11c2VybmFtZTpzb21lLXBhc3N3b3Jk
```

Invoking the function without the above header will respond with `401 Authorization Required`

<a id="canary-function"></a>
## Canary Function

You can add a canary function to an api gateway by adding another upstream to the api gateway.

> You can control the percentage of traffic that goes to the canary function by changing the percentage of the upstream.

Reviewer: switch for relevancy

Author: switch what?..

Reviewer: switch lines, sorry :) line 135 contents should come first

# API Gateway

> #### In this document

Reviewer 2: In This Section

- [No Authentication](#none-auth)
    - [Create](#create-none)
    - [Invoke](#invoke-none)
- [Basic Authentication](#basic-auth)
    - [Create](#create-basic)
    - [Invoke](#invoke-basic)
- [Canary Function](#canary-function)

<a id="none-auth"></a>

> ## No Authentication

Reviewer 2: authentication with lower case "a"

<a id="create-none"></a>

> ### Create API Gateways

Reviewer 2: ### Create API gateways

> You can create an api gateway with basic authentication by sending a POST request to the following endpoint:

Reviewer 2: **API** gateway

```
<nuclio-host-name>/api/api_gateways
```
With the following request body:
```json
{
    "spec": {
        "name": "<apigateway-name>",
        "description": "some description",
        "path": "/some/path", 
        "authenticationMode": "none",
        "upstreams": [
            {
                "kind": "nucliofunction",
                "nucliofunction": {
                    "name": "function-name-to-invoke"
                },
                "percentage": 0
            }
        ],
        "host": "<apigateway-name>-<project-name>.<nuclio-host-name>"
    },
    "metadata": {
        "labels": {
            "nuclio.io/project-name": "default"
        },
        "name": "<apigateway-name>"
    }
}
```

<a id="invoke-none"></a>

> ### Invoke API Gateways

Reviewer 2: **g**ateways

> To invoke it, simply send a request to the created API Gateway ingress (e.g.: `<apigateway-name>-<project-name>.<nuclio-host-name>`,

Reviewer 2: To invoke it, send a request to...

> specified on the request body `spec.host` )

Reviewer 2: specified on the request body `spec.host` )**.**
period at end

<a id="basic-auth"></a>

> ## Basic Authentication

Reviewer 2: **a**uthentication

> You may protect your function by setting the API Gateway with [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication). 

Reviewer 2: You can protect your function by applying [basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication) to the API Gateway.

> That means, to access a function, client would need to provide both username & password.

Reviewer 2: With basic authentication the client needs to provide both a username and password to access a function.

<a id="create-basic"></a>

> ### Create API Gateways

Reviewer 2: gateways

> You can create an api gateway with basic authentication by sending a POST request to the following endpoint:

Reviewer 2: API

                    "name": "function-name-to-invoke"
                },
                "percentage": 0
            }
        ],
        "host": "<apigateway-name>-<project-name>.<nuclio-host-name>",
        "authentication": {
            "basicAuth": {
                "username": "some-username",
                "password": "some-password"
            }
        }
    },
    "metadata": {
        "labels": {
            "nuclio.io/project-name": "default"
        },
        "name": "<apigateway-name>"
    }
}
```

<a id="invoke-basic"></a>
### Invoke API Gateways

To invoke it, simply send a request to the created API Gateway ingress (e.g.: `<apigateway-name>-<project-name>.<nuclio-host-name>`,
specified on the request body `spec.host`) with the following header:
```
key: Authorization
value: Basic base64encode("username:password")
```

*Example*:

For the following credentials:
```
"username": "some-username"
"password": "some-password"
```
base64 encoding of "some-username:some-password" is `c29tZS11c2VybmFtZTpzb21lLXBhc3N3b3Jk`(`echo "some-username:some-password" | base64 -d`), 
so the resulting header is:

```
Authorization: Basic c29tZS11c2VybmFtZTpzb21lLXBhc3N3b3Jk
```

> Invoking the function without the above header will respond with `401 Authorization Required`

Reviewer 2: Invoking the function without the above header results in`401 Authorization Required`

<a id="canary-function"></a>

> ## Canary Function

Reviewer 2: function

You can control the percentage of traffic that goes to a canary function by changing the percentage of the upstream.

> Add a canary function to an api gateway by adding another upstream to the api gateway, and set its `"percentage"` to a value between 0 and 100.

Reviewer 2: API and API

Make sure to set the percentage of the first function accordingly.

> 

Reviewer 2: specify the following:

> For instance, if you have two functions, `function-1` and `function-2`, and you want to have 80% of the traffic go to `function-1` and 20% of the traffic go to `function-2`, you can do the following:

Reviewer 2: For instance, if you have two functions, `function-1` and `function-2`, and you want to have 80% of the traffic go to `function-1` and 20% of the traffic go to `function-2`, specify the following:
